frontend/app/src/components/Sidebar/SidebarNav.tsx

```diff
@@ -102,60 +103,70 @@ const SidebarNav = ({
     ]
   )
 
-  let contents = null
+  let contents: ReactNode[] = []
+  const totalPages = appPages.length
+  const shouldShowViewButton =
+    hasSidebarElements && totalPages >= COLLAPSE_THRESHOLD
   if (navSections.length > 0) {
     const pagesBySectionHeader = groupBy(
       appPages,
       page => page.sectionHeader || ""
     )
     // For MPAv2: renders each NavSection with its respective header
-    contents = navSections.map(header => {
-      return (
+    let currentPageCount = 0
+    navSections.forEach(header => {
+      const sectionPages = pagesBySectionHeader[header] ?? []
+      let viewablePages = sectionPages
+
+      if (shouldShowViewButton && !expanded) {
+        if (currentPageCount >= NUM_PAGES_TO_SHOW_WHEN_COLLAPSED) {
+          // We cannot even show the section
+          return
+        } else if (
+          currentPageCount + sectionPages.length >
+          NUM_PAGES_TO_SHOW_WHEN_COLLAPSED
+        ) {
+          // We can partially show the section
+          viewablePages = sectionPages.slice(
+            0,
+            NUM_PAGES_TO_SHOW_WHEN_COLLAPSED - currentPageCount
+          )
+        }
+      }
+      currentPageCount += viewablePages.length
+
+      contents.push(
         <NavSection key={header} header={header}>
-          {(pagesBySectionHeader[header] ?? []).map(generateNavLinks)}
+          {viewablePages.map(generateNavLinks)}
         </NavSection>
       )
     })
   } else {
+    let viewablePages = appPages
+    if (shouldShowViewButton && !expanded) {
+      viewablePages = appPages.slice(0, NUM_PAGES_TO_SHOW_WHEN_COLLAPSED)
+    }
     // For MPAv1: single NavSection with all pages displayed
```

Comment:
Nit: Should probably update this comment, since I think it can be either strategy
```
// For MPAv1 / MPAv2 with no section headers, single NavSection with all pages
```